crates/turborepo-lib/src/globwatcher/mod.rs

```diff
@@ -0,0 +1,206 @@
+use std::{
+    collections::{HashMap, HashSet},
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+};
+
+use futures::StreamExt;
+use globwatch::{GlobSender, GlobWatcher, StopToken, Watcher};
+use itertools::Itertools;
+use log::{trace, warn};
+use notify::RecommendedWatcher;
+
+/// Tracks changes for a given hash. A hash is a unique identifier for a set of
+/// files. Given a hash and a set of globs to track, this will watch for file
+/// changes and allow the user to query for changes.
+#[derive(Clone)]
+pub struct HashGlobWatcher<T: Watcher> {
+    hash_globs: Arc<Mutex<HashMap<Arc<String>, Glob>>>,
+    glob_statuses: Arc<Mutex<HashMap<Arc<String>, HashSet<Arc<String>>>>>,
+    watcher: Arc<Mutex<Option<GlobWatcher<T>>>>,
+    config: GlobSender,
+}
+
+#[derive(Clone)]
+pub struct Glob {
+    include: HashSet<Arc<String>>,
+    exclude: HashSet<Arc<String>>,
+}
+
+impl HashGlobWatcher<RecommendedWatcher> {
+    pub fn new(flush_folder: PathBuf) -> Result<Self, globwatch::Error> {
+        let (watcher, config) = GlobWatcher::new(flush_folder)?;
+        Ok(Self {
+            hash_globs: Default::default(),
+            glob_statuses: Default::default(),
+            watcher: Arc::new(Mutex::new(Some(watcher))),
+            config,
+        })
+    }
+}
+
+impl<T: Watcher> HashGlobWatcher<T> {
+    /// Watches a given path, using the flush_folder as temporary storage to
+    /// make sure that file events are handled in the appropriate order.
+    pub async fn watch(&self, root_folder: PathBuf, token: StopToken) {
+        let start_globs = {
+            let lock = self.hash_globs.lock().expect("no panic");
+            lock.iter()
+                .flat_map(|(_, g)| &g.include)
+                .cloned()
+                .collect::<Vec<_>>()
+        };
+
+        let mut stream = match self.watcher.lock().expect("no panic").take() {
+            Some(watcher) => watcher.into_stream(token),
+            None => {
+                warn!("watcher already consumed");
+                return;
+            }
+        };
+
+        // watch all the globs currently in the map
+        for glob in start_globs {
+            self.config.include(glob.to_string()).await.unwrap();
```

Comment:
Can this fail? What happens if we fail to start watching a glob?

Reply:
I have reworked it to bubble up